include/seqan3/alignment/pairwise/policy/banded_score_dp_matrix_policy.hpp

```diff
@@ -0,0 +1,216 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::banded_score_dp_matrix_policy.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <limits>
+#include <memory>
+
+#include <range/v3/view/slice.hpp>
+#include <range/v3/view/zip.hpp>
+
+#include <seqan3/alignment/pairwise/policy/unbanded_dp_matrix_policy.hpp>
+#include <seqan3/range/shortcuts.hpp>
+#include <seqan3/std/ranges>
+#include <seqan3/std/span.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief A policy to allocate and manage a banded scoring matrix.
+ * \ingroup alingment_policy
+ * \tparam derived_t      The type of the derived class.
+ * \tparam allocator_type The type of the allocator used to allocate the score matrix.
+ */
+template <typename derived_t, typename allocator_type>
+class banded_score_dp_matrix_policy :
+    public unbanded_dp_matrix_policy<banded_score_dp_matrix_policy<derived_t, allocator_type>,  allocator_type>
+{
+private:
+
+    //!\brief The type of the base.
+    using base_t = unbanded_dp_matrix_policy<banded_score_dp_matrix_policy<derived_t, allocator_type>,  allocator_type>;
+
+    //!\brief Befriend CRTP derived type.
+    friend derived_t;
+
+    /*!\name Member types.
+     * \{
+     */
+    //!\brief The type of a matrix cell; inherited from `base_t`.
+    using cell_type = typename base_t::cell_type;
+    //!\brief The type of the score matrix; inherited from `base_t`.
+    using score_matrix_type = typename base_t::score_matrix_type;
+    //!\}
+
+    //!\brief A constant value for simulating minus infinity
+    static constexpr std::tuple_element_t<0, cell_type> INF =
+        std::numeric_limits<std::tuple_element_t<0, cell_type>>::lowest() / 2;
```

Comment:
why lowest / 2 ?

Reply:
to protect against underflow.

Reply:
you are decreasing the INF and therefore need more range? What if you use small ints and need the range?

Reply:
That's the problem of the algorithm. But for the scalar mode, this will actually never happen, as 32 bit integer operations will be the most efficient operations on most of our target cpus, so it will always use this. And in the case of SIMD I can use saturated arithmetics so there will be differences.

Reply:
And additionally, the value is never propagated through the matrix. It is only to make the banded code easier to implement. A better approximation would be lowest - gap_score. But this affects the interface of some functions and as explained above I don‘t see the use case as in scalar mode you won‘t compute with 8bit scores only and hope for a better performance.

Reply:
:+1: